crates/bevy_reflect/src/type_registry.rs

```diff
@@ -283,36 +292,40 @@ impl TypeRegistryArc {
 /// an implementation of the [`GetTypeRegistration`] trait.
 ///
 /// Along with the type's [`TypeInfo`] and [short name],
-/// this struct also contains a type's registered [`TypeData`].
+/// this struct also contains a type's registered [`BaseTypeData`].
 ///
 /// See the [crate-level documentation] for more information on type registration.
 ///
 /// # Example
 ///
 /// ```
-/// # use bevy_reflect::{TypeRegistration, std_traits::ReflectDefault, FromType};
-/// let mut registration = TypeRegistration::of::<Option<String>>();
+/// # use bevy_reflect::{TypeRegistration, std_traits::ReflectDefault};
+/// let registration = TypeRegistration::of::<Option<String>>();
 ///
 /// assert_eq!("core::option::Option<alloc::string::String>", registration.type_name());
 /// assert_eq!("Option<String>", registration.short_name());
 ///
-/// registration.insert::<ReflectDefault>(FromType::<Option<String>>::from_type());
+/// let registration = registration.register::<Option<String>, ReflectDefault>();
 /// assert!(registration.data::<ReflectDefault>().is_some())
 /// ```
 ///
 /// [short name]: bevy_utils::get_short_name
 /// [crate-level documentation]: crate
 pub struct TypeRegistration {
     short_name: String,
-    data: HashMap<TypeId, Box<dyn TypeData>>,
+    data: HashMap<TypeId, Box<dyn BaseTypeData>>,
     type_info: &'static TypeInfo,
+    /// A queue of callbacks to dispatch whenever this registration is registered
+    /// into a [`TypeRegistry`].
+    register_queue: Vec<fn(&mut TypeRegistry)>,
```

Comment:
Why store each `on_register` callbacks into the registration?

Wouldn't it be possible to:
1. amend the `Reflect` macro to call `.on_register()` with the reflected traits
2. amend `TypeRegistry::register_type_data` to call it?

Now we don't need this awkward handling of callbacks.

I don't like at all this solution. It makes very difficult to trace execution of code:
- How many times and when is this callback ran?
- Several callbacks on the same registration can behave in very confusing ways

Reply:
Yeah this is the part I was struggling with because it definitely is super awkward. We could definitely just add it to those two places. It just means that `on_register` isn't "guaranteed" to be called (which might be something we just accept).

And the reason for that is because a user could do `registry.get_mut(type_id).unwrap().insert::<ReflectFoo>(FromType::<MyStruct>::from_type())` without realizing they need to call the `on_register` function afterwards.

Reply:
@soqb do you have any thoughts on this? I think I agree with @nicopap that the best solution might be to just modify the callers rather than the registration API. It's probably not often someone is manually editing the `TypeRegistration` directly anyways, and we can leave a note in the documentation that `on_register` should be called for `TypeData`.

Reply:
Hm, I was beginning to implement this when I ran into a problem: `GetTypeRegistration` doesn't have access to the registry (so we can't call this function in the `Reflect` derive). We could add it as a parameter, but that starts to encroach on the work done in #5781.

Maybe this needs to happen after that PR gets merged?

Reply:
i don't think moving the `on_register` calls into the derived macro is really better. it's absolutely conceivable that someone would build their own `TypeRegistration` and would not think to call `on_register` because that's not how these kinds of api tend to work.

however it's also true that backtraces and quite possibly performance are left on the table because of the dynamic way we call the hooks.

i think the cleanest solution would involve simplifying the `GetTypeRegistration` trait with a total redesign like the following:

```rust
trait Registerable {
    fn register(registry: &mut TypeRegistry);
}
```

but that's a pretty sweeping change that should probably be left until both this PR and the recursive registration implementation, so i'm fine to leave this PR's implementation the way it was initially implemented, @MrGVSV

Reply:
> i don't think moving the `on_register` calls into the derived macro is really better. it's absolutely conceivable that someone would build their own `TypeRegistration` and would not think to call `on_register` because that's not how these kinds of api tend to work.

Yeah this was my original concern and why I made `insert` take `self`, so as to prevent users from accidentally not calling `on_register` after registration.

> think the cleanest solution would involve simplifying the `GetTypeRegistration` trait with a total redesign like the following:
> 
> ```rust
> trait Registerable {
>     fn register(registry: &mut TypeRegistry);
> }
> ```

Yeah, I generally agree. The recursive registration PR actually adds something like this. It keeps the current `get_type_registration` method, but also adds a `fn register_type_dependencies(registry: &mut TypeRegistry)` method.

If we wanted to move the `on_register` call into the `GetTypeRegistration` impl, that would probably be a good place to do so.

But again, this means we leave the API in a position where users could register type data without calling `on_register`. While I think that's okay, perhaps a better solution would be rethinking how we create, manage, and expose a `TypeRegistration` in the first place.